### PR TITLE
[ARCTIC-1208] Optimizing of native Iceberg table supports optimizing a part of partitions at a time

### LINKED ISSUE #1208: [Improvement]: Self-Optimizing for Native Iceberg tables should limit the file cnt for each Optimizing
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

When there are many files(like 100,000+) to be optimized for Native Iceberg tables, some performance issues will occur if we optimize all these files at a time:
 - write tasks and files into sysdb costs too much time, we have observed writing 600,000+ into sysdb costs 1 hour, which brings a time-consuming plan process
 - delete tasks and files from sysdb with a too-large transaction, exceed the limit of MySQL
 - too many tasks cost too much time to execute, which can't be committed for a long time

So, it's time and of great value to solve these problems.

### How should we improve?

We should introduce a table property to limit the files which the optimizing can handle at a time. 

For the partitioned table, we should optimize the partition one by one in the order of small file count until it reaches the file count limit for the optimizing.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by wangtaohz**
## Why are the changes needed?

fix #1208 

## Brief change log

  - Support only optimizing a part of partitions for Minor/Full Optimizing of native Iceberg table
  - when optimizing, partitions should be sorted by priority, and collect tasks one by one
  - for the native Iceberg table's Minor Optimizing, the priority should be the small-files count in the partition
  - for the native Iceberg table's Full Optimizing, the priority should be the delete-file size in the partition

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by wangtaohz** [comment]
Now, all files in a partition are optimized at a time, only optimizing a part of files in a partition is not supported, in consideration of complexity and risks.

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/AbstractArcticOptimizePlan.java [hunk 1]
```diff
@@ -84,12 +82,16 @@ public AbstractArcticOptimizePlan(ArcticTable arcticTable, TableOptimizeRuntime
...
     this.baseFileScanTasks = baseFileScanTasks;
     this.changeFiles = changeFiles;
     this.isCustomizeDir = false;
     this.currentChangeSnapshotId = changeSnapshotId;
-    this.currentBaseSnapshotId = baseSnapshotId;
+  }
+
+  @Override
+  protected boolean limitFileCnt() {
+    return false;
```
Comment (zhoujinsong) [inline ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/AbstractArcticOptimizePlan.java:94]: Why do the mixed format tables optimizing not limit file count?

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/AbstractOptimizePlan.java [hunk 2]
```diff
@@ -72,41 +78,77 @@ public OptimizePlanResult plan() {
...
 
-    List<BasicOptimizeTask> tasks = collectTasks(currentPartitions);
+    List<BasicOptimizeTask> tasks = collectTasks(getPartitionsToOptimizeInOrder());
 
     long endTime = System.nanoTime();
-    LOG.debug("{} ==== {} plan tasks cost {} ns, {} ms", tableId(), getOptimizeType(), endTime - startTime,
+    LOG.info("{} ==== {} plan tasks cost {} ns, {} ms", tableId(), getOptimizeType(), endTime - startTime,
         (endTime - startTime) / 1_000_000);
-    LOG.debug("{} {} plan get {} tasks", tableId(), getOptimizeType(), tasks.size());
+    LOG.info("{} {} plan get {} tasks", tableId(), getOptimizeType(), tasks.size());
```
[the remark this answers is not in the document]
Comment (wangtaohz) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/AbstractOptimizePlan.java:86]: OK, I will combine them

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/AbstractArcticOptimizePlan.java (hunk 1 above)
Comment (zhoujinsong) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/AbstractArcticOptimizePlan.java:94]: AFAIK, the partition weight for mixed format table optimizing may not be a simple number, am I right?
If yes, the current `PartitionWeight` definition may not be good for scaling.

Comment (zhoujinsong) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/AbstractArcticOptimizePlan.java:94]: We can support mixed-format tables for this feature in another PR, but we may create an issue to address that.

Comment (wangtaohz) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/AbstractArcticOptimizePlan.java:94]: OK, I create another issue to address the mix-format table. And I will abstract an interface `Weight` to replace a simple long.

Comment (wangtaohz) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/AbstractArcticOptimizePlan.java:94]: related issue https://github.com/NetEase/arctic/issues/1213

**Review by zhoujinsong — COMMENTED** [review]
LGTM.

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Comment by XBaith** [comment]
@wangtaohz Good idea! Optimizing tables by partition is an effective way to merge files, but I think it's still a headache for unpartitioned tables. Would you consider improving the optimization of unpartitioned tables in the future?

**Comment by wangtaohz** [comment]
> Would you consider improving the optimization of unpartitioned tables in the future?

Yes, It is going to be optimized in the future. If you have cases that unpartitioned tables containing many small files now, I will raise the priority.

**Comment by XBaith** [comment]
>

In our production situation, we have not encountered this issue, so just follow your plan.

**Pull request merged**